### ISSUE TITLE
GPlus native integration

### ISSUE THREAD

**Issue by joaom182** [auth0/Lock.Android#45]
I'm trying to integrate native authentication with GPlus, but is not working to me.

* I set up my application in Auth0 with the secrets and tokens needed for integration with Google

* I follow the First Step ([https://developers.google.com/+/mobile/android/getting-started](https://developers.google.com/+/mobile/android/getting-started))

* Added library dependencies on my gradle.file

```
compile 'com.auth0.android:lock:1.1.+'
compile 'com.auth0.android:lock-googleplus:1.1.+'
```

* Added proGuardFiles on gradle file

```
buildTypes {
  release {
    minifyEnabled true
    proguardFile '../proguard/proguard-google-play-services.pro' //G+ native auth
    proguardFile '../proguard/proguard-android-async-http.pro' //Auth0 core
    proguardFile '../proguard/proguard-jackson-2.pro' //Auth0 core
    proguardFile '../proguard/proguard-square-otto.pro' //Lock
    proguardFile '../proguard/proguard-lock.pro' //Lock
    //Add your app's specific proguard rules
  }  
```

* I implemented the GoogleApiClient connection and Lock scaffolding on my Activity

``` java
// WelcomeActivity.java
package com.squidit.activities;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.view.View;

import com.auth0.core.Token;
import com.auth0.core.UserProfile;
import com.auth0.lock.Lock;
import com.auth0.lock.LockActivity;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.plus.Plus;
import com.squidit.fragments.LoginSlideFragment;
import com.squidit.squid.R;

import java.util.ArrayList;


public class WelcomeActivity extends FragmentActivity {

    private GoogleApiClient mGoogleApiClient;
    private LocalBroadcastManager broadcastManager;
    private BroadcastReceiver authenticationReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            UserProfile profile = intent.getParcelableExtra("profile");
            Token token = intent.getParcelableExtra("token");
            Log.i("LockLog", "User " + profile.getName() + " logged in");
            goToMainActivity();
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome);

        Plus.PlusOptions options = new Plus.PlusOptions.Builder()
                .addActivityTypes("http://schemas.google.com/AddActivity",
                        "http://schemas.google.com/ReviewActivity")
                .build();

        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addApi(Plus.API, options)
                .addScope(Plus.SCOPE_PLUS_LOGIN)
                .build();

        broadcastManager = LocalBroadcastManager.getInstance(this);
        broadcastManager.registerReceiver(authenticationReceiver, new IntentFilter(Lock.AUTHENTICATION_ACTION));
    }

    @Override
    protected void onStart() {
        super.onStart();
        mGoogleApiClient.connect();
    }

    @Override
    protected void onStop() {
        mGoogleApiClient.disconnect();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        broadcastManager.unregisterReceiver(authenticationReceiver);
    }

    @Override
    public void onBackPressed() {
            super.onBackPressed();
    }

    private void goToMainActivity(){
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
    }

    public void enter(View view){
        Intent lockIntent = new Intent(this, LockActivity.class);
        startActivity(lockIntent);
    }

}
```

* Implemented an Application to initialize Lock

``` java
// SquidApplication.java
package com.squidit.application;

import android.app.Application;

import com.auth0.core.Strategies;
import com.auth0.facebook.FacebookIdentityProvider;
import com.auth0.googleplus.GooglePlusIdentityProvider;
import com.auth0.lock.Lock;
import com.auth0.lock.LockBuilder;
import com.auth0.lock.LockProvider;

public class SquidApplication extends Application implements LockProvider {

    private Lock lock;

    public void onCreate() {
        super.onCreate();
        lock = new LockBuilder()
                .loadFromApplication(this)
                .closable(true)
                .build();

        FacebookIdentityProvider facebook = new FacebookIdentityProvider();
        GooglePlusIdentityProvider googleplus = new GooglePlusIdentityProvider(this);

        lock.setProvider(Strategies.GooglePlus.getName(), googleplus);
        lock.setProvider(Strategies.Facebook.getName(), facebook);
    }

    @Override
    public Lock getLock() {
        return lock;
    }
}
```

* Configured Manifest.xml

``` xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    package="com.squidit.squid">

    <!-- To auto-complete the email text field in the login form with the user's emails -->
    <uses-permission android:name="android.permission.INTERNET"/>
    <uses-permission android:name="android.permission.GET_ACCOUNTS" />
    <uses-permission android:name="android.permission.READ_PROFILE" />
    <uses-permission android:name="android.permission.READ_CONTACTS" />
    <uses-permission android:name="android.permission.USE_CREDENTIALS" />
    <meta-data android:name="com.google.android.gms.version" android:value="@integer/google_play_services_version" />

    <application
        android:name="com.squidit.application.SquidApplication"
        android:allowBackup="true"
        android:icon="@drawable/ic_squid"
        android:label="@string/app_name"
        android:theme="@style/DefaultTheme"
        tools:replace="android:icon">
        <activity
            android:name="com.squidit.activities.WelcomeActivity"
            android:label="@string/app_name"
            android:theme="@style/DefaultTheme" >
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
        <!--Auth0 Lock-->
        <activity
            android:name="com.auth0.lock.LockActivity"
            android:theme="@style/Lock.Theme"
            android:screenOrientation="portrait"
            android:launchMode="singleTask">
            <intent-filter>
                <action android:name="android.intent.action.VIEW"/>
                <category android:name="android.intent.category.DEFAULT"/>
                <category android:name="android.intent.category.BROWSABLE"/>
                <data android:scheme="a0MY_AUTH0_CLIENT_ID" android:host="MY_AUTH0_DOMAIN.auth0.com"/>
            </intent-filter>
        </activity>
        <meta-data android:name="com.auth0.lock.client-id" android:value="@string/auth0_client_id"/>
        <meta-data android:name="com.auth0.lock.tenant" android:value="@string/auth0_tenant_name"/>
        <!--Auth0 Lock End-->
        <activity
            android:name="com.squidit.activities.MainActivity"
            android:label="@string/title_activity_main" >
        </activity>
        <activity android:name="com.facebook.LoginActivity"/>
        <meta-data android:name="com.facebook.sdk.ApplicationId" android:value="@string/facebook_app_id"/>
    </application>

</manifest>
```

Buuut, throws an exception in Android Studio when i try to login with GPlus:

```
com.auth0.googleplus.GooglePlusIdentityProvider: Connection failed with unrecoverable error
com.auth0.lock.LockActivity: Failed to authenticate user
com.auth0.googleplus.GooglePlusIdentityProvider: Failed to clear G+ Session.
java.lang.IllegalStateException: GoogleApiClient must be connected
```

And on the android application:

![screenshot_2015-02-18-22-02-13](https://cloud.githubusercontent.com/assets/3612098/6260346/ac19bfd6-b7c4-11e4-8508-e16bae3758de.png)

What I'm doing wrong?

**Comment by hzalaz** [comment]
@joaom182 I'll check it out. Just one question, did you follow the first step [here](https://developers.google.com/+/mobile/android/getting-started)?

**Comment by joaom182** [comment]
Yeah, see the Screenshot of my google console

![2015-02-19_120017](https://cloud.githubusercontent.com/assets/3612098/6267933/99cb997e-b82f-11e4-809f-05e0b673206f.jpg)

**Comment by joaom182** [comment]
Just one question... is it really necessary I implement the GoogleApiClient connection?

``` java
mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addApi(Plus.API, options)
                .addScope(Plus.SCOPE_PLUS_LOGIN)
                .build();
```

or the Lock package will implement this internally?

**Comment by hzalaz** [comment]
Internally Lock uses `GoogleApiClient` to authenticate with G+, so if you don't need to retrieve more information later you can remove it. I've checked the code and that error is reported when G+ raises an exception during auth, most of the time without a helpful message but the cause is that some configuration is missing for G+ integration. Unfortunately I don't see whats missing in your code so I'll try to reproduce it in a new project.

**Comment by joaom182** [comment]
I'm suspecting the package name, in my Manifest.xml is `com.squidit.squid` and on the GPlus console is `com.squidit`, i will try investigate this, thanks a lot for you help.

**Comment by hzalaz** [comment]
Sure, you can find me too in [chat.auth0.com](https://chat.auth0.com). 
BTW, one last thing:
```xml
<meta-data android:name="com.google.android.gms.version" android:value="@integer/google_play_services_version" />
```
should be inside `<application>`

I'll release a new version this week, and I'll add a little more logs for that exception.

**Comment by joaom182** [comment]
* Fixed the meta-data tag on Manifest.xml
* Fixed package name on Google Console

But is still throwing exception

**Comment by pmq** [comment]
Same thing here. @hzalaz could it be that Lock.Android depends on `com.google.android.gms:play-services-plus:6.5.87` but most of us are using the current `7.0.0` version ?

**Comment by pmq** [comment]
Some followup after testing on LG Nexus 5 / Android 5.1 and Samsung Galaxy S5 / Android 4.4.4. In fact, in my case, I have exactly the same stacktrace as @joaom182, that is for completeness :
```
05-24 10:54:26.615    8808-8808/__REDACTED__ E/ViewRootImpl﹕ sendUserActionEvent() mView == null
05-24 10:54:26.775    8808-8808/__REDACTED__ E/com.auth0.googleplus.GooglePlusIdentityProvider﹕ Connection failed with unrecoverable error
05-24 10:54:26.775    8808-8808/__REDACTED__ E/com.auth0.lock.LockActivity﹕ Failed to authenticate user
05-24 10:54:26.785    8808-8808/__REDACTED__ E/com.auth0.googleplus.GooglePlusIdentityProvider﹕ Failed to clear G+ Session
    java.lang.IllegalStateException: GoogleApiClient must be connected.
            at com.google.android.gms.common.internal.zzx.zza(Unknown Source)
            at com.google.android.gms.plus.Plus.zzf(Unknown Source)
            at com.google.android.gms.internal.zztm.clearDefaultAccount(Unknown Source)
            at com.auth0.googleplus.GooglePlusIdentityProvider.clearSession(GooglePlusIdentityProvider.java:108)
            at com.auth0.lock.LockActivity.onAuthenticationError(LockActivity.java:221)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.squareup.otto.EventHandler.handleEvent(EventHandler.java:89)
            at com.squareup.otto.Bus.dispatch(Bus.java:384)
            at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:367)
            at com.squareup.otto.Bus.post(Bus.java:336)
            at com.auth0.lock.identity.LockIdentityProviderCallback.onFailure(LockIdentityProviderCallback.java:62)
            at com.auth0.googleplus.GooglePlusIdentityProvider.onConnectionFailed(GooglePlusIdentityProvider.java:147)
            at com.google.android.gms.common.internal.zzl.zzg(Unknown Source)
            at com.google.android.gms.common.api.zzd.zza(Unknown Source)
            at com.google.android.gms.common.api.zzd.zzie(Unknown Source)
            at com.google.android.gms.common.api.zzd.zzc(Unknown Source)
            at com.google.android.gms.common.api.zzd$9.onConnectionFailed(Unknown Source)
            at com.google.android.gms.common.internal.zzl.zzg(Unknown Source)
            at com.google.android.gms.common.internal.zzk$zza.zzf(Unknown Source)
            at com.google.android.gms.common.internal.zzk$zza.zzc(Unknown Source)
            at com.google.android.gms.common.internal.zzk$zza.zzi(Unknown Source)
            at com.google.android.gms.common.internal.zzk$zzc.zzjg(Unknown Source)
            at com.google.android.gms.common.internal.zzk$zzb.handleMessage(Unknown Source)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:146)
            at android.app.ActivityThread.main(ActivityThread.java:5679)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1291)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1107)
            at dalvik.system.NativeStart.main(Native Method)
```

But... if I click a second time on the G+ button after dismissing the 'There was an error during authentication', it now works.

Let me know if you need me to test a new build.

**Comment by hzalaz** [comment]
Should be fixed in version `2.0.0` of `lock-googleplus`
